Source/ThirdParty/ANGLE/src/tests/gl_tests/VariableRasterizationRateTestMetal.mm

```diff
@@ -0,0 +1,300 @@
+    }
+
+  private:
+    void release()
+    {
+#if !__has_feature(objc_arc)
+        [mMap release];
+#endif
+    }
+
+    id<MTLRasterizationRateMap> mMap = nil;
+};
+
+ScopeMetalRasterizationRateMapRef CreateMetalRasterizationRateMap(id<MTLDevice> deviceMtl,
+                                                                  int width,
+                                                                  int height,
+                                                                  int horizontalCount,
+                                                                  const float *horizontal,
+                                                                  int verticalCount,
+                                                                  const float *vertical)
+{
+    @autoreleasepool
+    {
+        MTLRasterizationRateLayerDescriptor *rrLayer = [[MTLRasterizationRateLayerDescriptor alloc]
+            initWithSampleCount:MTLSizeMake(horizontalCount, verticalCount, 0)
+                     horizontal:horizontal
+                       vertical:vertical];
+        MTLRasterizationRateMapDescriptor *rrDesc    = [MTLRasterizationRateMapDescriptor
+            rasterizationRateMapDescriptorWithScreenSize:MTLSizeMake(width, height, 0)
+                                                   layer:rrLayer];
+        rrDesc.label                                 = @"Test RasterizationRate";
+
+        ScopeMetalRasterizationRateMapRef rateMap(
+            [deviceMtl newRasterizationRateMapWithDescriptor:rrDesc]);
+        return rateMap;
+    }
+}
+
+class VariableRasterizationRateTestMetal : public ANGLETest<>
+{
+  protected:
+    VariableRasterizationRateTestMetal()
+    {
+        setWindowWidth(kWindowWidth);
+        setWindowHeight(kWindowHeight);
+        setConfigRedBits(8);
+        setConfigGreenBits(8);
+        setConfigBlueBits(8);
+        setConfigAlphaBits(8);
+        setConfigDepthBits(24);
+    }
+
+    void testSetUp() override
+    {
+        ANGLE_UNUSED_VARIABLE(kFS);
+        mProgram = CompileProgram(kVS, kCheckerBoardFS);
+        if (!mProgram)
+        {
+            FAIL() << "shader compilation failed.";
+        }
+        mUniformResolutionLocation = glGetUniformLocation(mProgram, "u_resolution");
+        mUniformCheckCountLocation = glGetUniformLocation(mProgram, "u_checkCount");
+
+        ASSERT_GL_NO_ERROR();
+    }
+
+    void testTearDown() override {}
+
+    id<MTLDevice> getMtlDevice()
+    {
+        EGLAttrib angleDevice = 0;
+        EGLAttrib device      = 0;
+        EXPECT_EGL_TRUE(
+            eglQueryDisplayAttribEXT(getEGLWindow()->getDisplay(), EGL_DEVICE_EXT, &angleDevice));
+
+        EXPECT_EGL_TRUE(eglQueryDeviceAttribEXT(reinterpret_cast<EGLDeviceEXT>(angleDevice),
+                                                EGL_METAL_DEVICE_ANGLE, &device));
+
+        return (__bridge id<MTLDevice>)reinterpret_cast<void *>(device);
+    }
+
+    template <size_t N, size_t M>
```

Comment:
Might suggest X or W for the horizontal direction and Y or H for the vertical direction, just for ease of reading.